data-prepper-plugins/http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceTest.java

```diff
@@ -96,10 +100,62 @@ public void testHTTPJsonResponse() {
                         .build(),
                 HttpData.ofUtf8("[{\"log\": \"somelog\"}]"))
                 .aggregate()
-                .whenComplete((i, ex) -> assertThat(i.status().code()).isEqualTo(200)).join();
+                .whenComplete((i, ex) -> assertThat(i.status()).isEqualTo(HttpStatus.OK)).join();
         Assertions.assertFalse(testBuffer.isEmpty());
     }
 
+    @Test
+    public void testHTTPJsonResponse429() throws InterruptedException {
+        // Prepare
+        final Map<String, Object> settings = new HashMap<>();
+        final int testMaxPendingRequests = 1;
+        final int testThreadCount = 1;
+        final int clientTimeoutInMillis = 100;
+        final int serverTimeoutInMillis = (testMaxPendingRequests + testThreadCount + 1) * clientTimeoutInMillis;
+        settings.put(HTTPSourceConfig.REQUEST_TIMEOUT, serverTimeoutInMillis);
+        settings.put(HTTPSourceConfig.MAX_PENDING_REQUESTS, testMaxPendingRequests);
+        settings.put(HTTPSourceConfig.THREAD_COUNT, testThreadCount);
+        testPluginSetting = new PluginSetting("http", settings);
+        testPluginSetting.setPipelineName("pipeline");
+        HTTPSourceUnderTest = new HTTPSource(testPluginSetting);
+        // Start the source
+        HTTPSourceUnderTest.start(testBuffer);
+        final RequestHeaders testRequestHeaders = RequestHeaders.builder().scheme(SessionProtocol.HTTP)
+                .authority("127.0.0.1:2021")
+                .method(HttpMethod.POST)
+                .path("/log/ingest")
+                .contentType(MediaType.JSON_UTF_8)
+                .build();
+        final HttpData testHttpData = HttpData.ofUtf8("[{\"log\": \"somelog\"}]");
+
+        // Fill in the buffer
+        WebClient.of().execute(testRequestHeaders, testHttpData).aggregate().whenComplete(
+                        (response, ex) -> assertThat(response.status()).isEqualTo(HttpStatus.OK)).join();
+
+        // Send requests to throttle the server when buffer is full
+        // Set the client timeout to be less than source serverTimeoutInMillis / (testMaxPendingRequests + testThreadCount)
+        WebClient testWebClient = WebClient.builder().responseTimeoutMillis(clientTimeoutInMillis).build();
+        for (int i = 0; i < testMaxPendingRequests + testThreadCount; i++) {
+            try {
+                testWebClient.execute(testRequestHeaders, testHttpData).aggregate().join();
+            } catch (CompletionException e) {
+                assertThat(e.getCause()).isInstanceOf(ResponseTimeoutException.class);
```

Comment:
I should have noted this in the last PR, but this `assertThat` will not be called unless an exception is actually thrown.

Please change to:
```
CompletionException actualException = assertThrows(CompletionException.class, () -> testWebClient.execute(testRequestHeaders, testHttpData).aggregate().join());
assertThat(actualException.getCause()).isInstanceOf(ResponseTimeoutException.class);
```

Also, below.

Reply:
Great catch!